### PR TITLE
Version improvements

### PR DESCRIPTION
**Pull request by cproensa**
Some improvements for version related apis

- Clean up of queries and cache. Less database queries.
- Fix some issues due to wrong cache.
- Fix including inherited versions in filters

https://www.mantisbt.org/bugs/view.php?id=12261
https://www.mantisbt.org/bugs/view.php?id=22100
https://www.mantisbt.org/bugs/view.php?id=23245
https://www.mantisbt.org/bugs/view.php?id=24821

### CODE REVIEW

File: core/print_api.php [hunk 1]
```diff
@@ -846,34 +846,34 @@ function print_os_build_option_list( $p_os_build, $p_user_id = null ) {
...
 
 /**
- * Print the option list for versions
+ * Print the option list for versions.
+ * All versions related for each project will be printed. Those include, for each
+ * project, the directly linked versions and the inherited versions if applicable.
+ *
  * @param string  $p_version       The currently selected version.
- * @param integer $p_project_id    Project id, otherwise current project will be used.
+ * @param integer $p_project_ids   Array of project ids, of null to use current project.
```
Comment (dregad) [inline core/print_api.php]: PHPdoc type is wrong based on description, should be `array|null`

File: core/prepare_api.php [hunk 2]
```diff
@@ -94,20 +94,23 @@ function prepare_user_name( $p_user_id ) {
  *
  * @param integer $p_project_id The project id.
  * @param integer $p_version_id The version id.  If false then this method will return an empty string.
+ * @param boolean $p_show_project Force prepending project name to version
```
Comment (dregad) [inline core/prepare_api.php]: PHPDoc says boolean, but param defaults to _null_. Change PHPDoc type to `boolean|null` and document what null means, or change default param value to _false_ (whatever is appropriate)

File: core/print_api.php [hunk 3]
```diff
@@ -846,34 +846,34 @@ function print_os_build_option_list( $p_os_build, $p_user_id = null ) {
...
-function print_version_option_list( $p_version = '', $p_project_id = null, $p_released = null, $p_leading_blank = true, $p_with_subs = false ) {
-	if( null === $p_project_id ) {
-		$c_project_id = helper_get_current_project();
-	} else {
-		$c_project_id = (int)$p_project_id;
+function print_version_option_list( $p_version = '', $p_project_ids = null, $p_released = null, $p_leading_blank = true ) {
+	if( null === $p_project_ids ) {
+		$p_project_ids = helper_get_current_project();
 	}
+	$t_project_ids = is_array( $p_project_ids ) ? $p_project_ids : array( $p_project_ids );
```
Comment (vboctor) [inline core/print_api.php]: How about moving this conversion to array be moved inside `version_get_all_rows()`?

File: core/print_api.php [hunk 4]
```diff
@@ -846,34 +846,34 @@ function print_os_build_option_list( $p_os_build, $p_user_id = null ) {
...
 	# Note: Filter API specifies selected versions as an array
 	if( !is_array( $p_version ) ) {
 		if( !empty( $p_version ) ) {
-			$t_version_id = version_get_id( $p_version, $c_project_id );
-			if( $t_version_id !== false ) {
-				$t_versions[] = version_cache_row( $t_version_id );
+			foreach( $t_project_ids as $t_project_id ) {
+				$t_version_id = version_get_id( $p_version, $t_project_id );
+				if( $t_version_id !== false ) {
+					$t_versions[] = version_cache_row( $t_version_id );
```
Comment (vboctor) [inline core/print_api.php]: Shouldn't methods that are not part of `version_api.php` use a more public method than `version_cache_row()`? For example, `version_get()` or something along these lines?

File: core/print_api.php [hunk 5]
```diff
@@ -900,8 +902,8 @@ function print_version_option_list( $p_version = '', $p_project_id = null, $p_re
 			$t_listed[] = $t_version_version;
 			echo '<option value="' . $t_version_version . '"';
 			check_selected( $p_version, $t_version['version'] );
-
-			$t_version_string = string_attribute( prepare_version_string( $c_project_id, $t_version['id'] ) );
+			$c_project_id = $t_version['project_id'];
```
Comment (vboctor) [inline core/print_api.php]: Use `$t_project_id` instead of `$c_project_id`.

File: core/version_api.php [hunk 6]
```diff
@@ -202,6 +213,47 @@ function version_cache_row( $p_version_id, $p_trigger_errors = true ) {
...
+	}
+
+	$t_query = new DbQuery();
+	$t_sql = 'SELECT * FROM {project_version} WHERE ' . $t_query->sql_in( 'project_id', $t_ids_to_fetch );
+	$t_query->sql( $t_sql );
+	while( $t_row = $t_query->fetch() ) {
+		$c_project_id = (int)$t_row['project_id'];
+		$c_version_id = (int)$t_row['id'];
+		if( !isset( $g_cache_versions[$c_version_id] ) ) {
+			$g_cache_versions[$c_version_id] = $t_row;
```
Comment (vboctor) [inline core/version_api.php]: Why not refresh the cache anyway since the version was retrieved from the db?

File: core/version_api.php [hunk 7]
```diff
@@ -438,156 +490,70 @@ function version_remove_all( $p_project_id ) {
...
-		$g_cache_versions_project[(int)$t_project_id] = $t_row;
-	}
-}
-
 /**
- * Return all versions for the specified project
- * @param integer $p_project_id A valid project id.
+ * Return all versions for the specified project or projects list
+ * Returned versions are ordered by reverse 'date_order'
+ * @param integer|array $p_project_ids  A valid project id, or array of ids
```
Comment (vboctor) [inline core/version_api.php]: Looks like you already support a single version or an array, then doing the same from print_api is redundant.

File: core/print_api.php (hunk 3 above)
Comment (cproensa) [reply core/print_api.php]: $t_project_ids is used later in the method, and having it normalized to an array is cleaner.
`version_get_all_rows()` may be called from other places, so it also has to normalize the input separately.

File: core/print_api.php (hunk 4 above)
Comment (cproensa) [reply core/print_api.php]: `version_get()`returns a VersionData object. Most of the places we user directly version "rows", but the only method to get a row is `version_cache_row()`.
Your comment is in the line of why we have `bug_get_row()` which is a passthrough wrapper for `bug_cache_row()`

File: core/version_api.php (hunk 7 above)
Comment (cproensa) [reply core/version_api.php]: Is this related to previous comment?: https://github.com/mantisbt/mantisbt/pull/1465#discussion_r257476614
about the `is_array( $p_project_ids ) ? $p_project_ids : array( $p_project_ids )`

File: core/print_api.php (hunk 4 above)
Comment (vboctor) [reply core/print_api.php]: Makes sense to have `version_get_row()` then.  Not a blocker, but nice to do.

File: core/version_api.php (hunk 7 above)
Comment (vboctor) [reply core/version_api.php]: Yep, seems you have reasons for duplication, so I'm ok here.

File: core/print_api.php (hunk 3 above)
Comment (vboctor) [reply core/print_api.php]: ok

**Pull request merged**